### PR TITLE
Feature/new mongodb driver

### PR DESCRIPTION
**Pull request by tacxou**
Hello!

* Type: breaks bc | new feature
* Link to issue: https://github.com/phalcon/cphalcon/issues/13697

**In raising this pull request, I confirm the following:**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/phalcon/cphalcon/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist
- [x] I wrote some tests for this PR
- [ ] I have updated the relevant CHANGELOG
- [ ] I have created a PR for the [documentation](https://github.com/phalcon/docs) about this change

* Added `Phalcon\Mvc\Collection\Document::__construct()`
* Added `Phalcon\Mvc\Collection\Document::count()`
* Changed `Phalcon\Mvc\Collection\Document`  
  * Now implements `Countable`
* Changed `Phalcon\Mvc\Collection::findFirst()`
  * Now returns a `null` or `CollectionInterface` 
* Changed `Phalcon\Mvc\Collection::getId()`
  * Now returns a `\MongoDB\BSON\ObjectId` 
* Depreciated `Phalcon\Mvc\Collection::createIfNotExist()`
* Deleted `Phalcon\Mvc\Collection::summatory()`
  * This method to group records no longer exists in the new driver. It can be easily replaced by the MongoDB [$group](https://docs.mongodb.com/manual/reference/operator/aggregation/group/) in [aggregation framework](https://docs.mongodb.com/manual/aggregation/), you can use `Phalcon\Mvc\Collection::aggregate()` method. 
* Changed `Phalcon\Mvc\Collection\Manager::getCustomEventsManager()`
  * Now return a `null` or `Phalcon\Events\ManagerInterface`
* Changed `Phalcon\Mvc\Collection\Manager::getConnection()`
  * Now return a `\MongoDB\Database` or `Phalcon\Db\Adapter\AdapterInterface`
* Refactoring `Phalcon\Mvc\Collection\Manager`
  * Events triggered in collections are now prefixed with `collection:` instead of `model:`
  * Modification of variable names to match collections (`var model` to `var collection`)

- Refactoring `Phalcon\Mvc\Collection`
  - Modification of variables of type `MongoId` to the new type `ObjectId`
  - Modification of variable names to match collections (`var mongoId` to `var objectId`)

----

Thanks

### CODE REVIEW

**Comment by sergeyklay** [comment]
@tacxticx88 Could you please rebase

File: phalcon/Mvc/Collection.zep [hunk 1]
```diff
@@ -1689,4 +1644,46 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
 
         return collection;
     }
+
+    /**
+	 * {@inheritdoc}
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: Could you correct indents

File: phalcon/Mvc/Collection.zep [hunk 2]
```diff
@@ -1689,4 +1644,46 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
...
+		let this->modelsManager = manager;
+
+		/**
+		 * Update the objects attributes
+		 */
+		for key, value in data {
+			let this->{key} = value;
+		}
+
+		if (method_exists(this, "afterFetch")) {
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]:
```suggestion
		if method_exists(this, "afterFetch") {
```

But in general I wold like work with interfaces

File: phalcon/Mvc/Collection/ManagerInterface.zep [hunk 3]
```diff
@@ -41,56 +40,56 @@ interface ManagerInterface
...
      */
-    public function addBehavior(<CollectionInterface> model, <BehaviorInterface> behavior);
+    public function addBehavior(<CollectionInterface> collection, <BehaviorInterface> behavior);
 
     /**
-     * Returns the connection related to a model
+     * Returns the connection related to a collection
      */
-    public function getConnection(<CollectionInterface> model) -> <AdapterInterface>;
+    public function getConnection(<CollectionInterface> collection);
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection/ManagerInterface.zep]: what about return type?

File: phalcon/Mvc/Collection.zep [hunk 4]
```diff
@@ -1015,6 +1024,8 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
 
     /**
      * Allows to perform a summatory group for a column in the collection
+     *
+     * @deprecated
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: Please don't introduce deprecations in the brand new version

File: phalcon/Mvc/Collection.zep [hunk 5]
```diff
@@ -1194,24 +1205,18 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
...
-        if typeof status == "array" {
-            if fetch ok, status["ok"] {
-                if ok {
-                    let success = true;
-                }
-            }
-        } else {
-            let success = false;
+        if status->isAcknowledged() {
+            let success = true;
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: Otherwise NULL?

File: phalcon/Mvc/Collection.zep [hunk 6]
```diff
@@ -558,13 +560,13 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
      * }
      * ```
      */
-    public static function findById(var id) -> <CollectionInterface> | null
+    public static function findById(var id) -> <CollectionInterface> | bool
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: Could you please return this and all other similar changes (null -> bool) back.

**Review by sergeyklay — CHANGES_REQUESTED** [review]

File: phalcon/Mvc/Collection.zep (hunk 4 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: Okay, should I delete the methods?

File: phalcon/Mvc/Collection/ManagerInterface.zep (hunk 3 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection/ManagerInterface.zep]: getConnection returns a "MongoDB Database", do I have to correct that? In phalcon:3.4.4.x, getConnection returned "\MongoDB"

File: phalcon/Mvc/Collection.zep (hunk 6 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: the "findFirst" and "findById" must return a null? Because in the original code, the return type was "Collection" or "false" ?

File: phalcon/Mvc/Collection.zep (hunk 5 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: What do you mean?

File: phalcon/Mvc/Collection.zep (hunk 4 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: should I delete the methods in this case?

Comment (sergeyklay) [reply phalcon/Mvc/Collection.zep]: Yes, I think so. And don't forget to amend change log for _any notable change_, e.g. changed method signature,  method deletion, etc.

File: phalcon/Mvc/Collection.zep (hunk 5 above)
Comment (sergeyklay) [reply phalcon/Mvc/Collection.zep]: I mean that `success` is NULL before this line

File: phalcon/Mvc/Collection.zep (hunk 6 above)
Comment (sergeyklay) [reply phalcon/Mvc/Collection.zep]: Please make it consistent with https://github.com/phalcon/cphalcon/issues/14044

File: phalcon/Mvc/Collection/ManagerInterface.zep (hunk 3 above)
Comment (sergeyklay) [reply phalcon/Mvc/Collection/ManagerInterface.zep]: as a far as I see it should return an interface/abstract class instance, isn't?
```zep
        // ...
        let connection = container->getShared(service);
        // ...

        return connection;
```

Comment (tacxticx88) [reply phalcon/Mvc/Collection/ManagerInterface.zep]: I didn't change this part of the code, it returns an instance of \MongoDB\Database;  in 3.4.x it returned an instance of \Mongo

File: phalcon/Mvc/Collection.zep (hunk 4 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: It works, I'll do it. For the "createIfNotExists" method, is it useful knowing that it is not documented and that the "uniqness" validator replaces this behavior ?

File: phalcon/Mvc/Collection.zep (hunk 6 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: Do the methods have to lift an exception?

File: phalcon/Mvc/Collection/ManagerInterface.zep (hunk 3 above)
Comment (sergeyklay) [reply phalcon/Mvc/Collection/ManagerInterface.zep]: But now it isn't 3.4. Let's change method signature

File: phalcon/Mvc/Collection.zep (hunk 4 above)
Comment (niden) [reply phalcon/Mvc/Collection.zep]: You can either add that in the changelog or add it to the documents or even send me a message so that we can add it in the documentation. :)

File: phalcon/Mvc/Collection.zep (hunk 6 above)
Comment (niden) [reply phalcon/Mvc/Collection.zep]: The `findFirst` etc. have been changed to return `null` or the record itself. In this case it will be a Collection object. Have a look at the implementation that the model has and let us replicate that. 

These are the changes made to make the methods return `null`

https://github.com/phalcon/cphalcon/pull/14089/files

**Comment by niden** [comment]
@tacxticx88 when you have some time can you look at this PR:

https://github.com/phalcon/cphalcon/pull/13779/files

to make sure that the supplied fix is included with your PR?

Thanks

File: phalcon/Mvc/Collection.zep (hunk 6 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: Normally these changes are made now

File: phalcon/Mvc/Collection.zep [hunk 7]
```diff
@@ -32,7 +31,7 @@ use Serializable;
  * This component implements a high level abstraction for NoSQL databases which
  * works with documents
  */
-abstract class Collection implements EntityInterface, CollectionInterface, InjectionAwareInterface, Serializable
+abstract class Collection implements EntityInterface, CollectionInterface, InjectionAwareInterface, Unserializable, \Serializable
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: Why do we need implement Unserializable? Can we avoid this?

Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: It is this interface and the bsonUnserialize method that allows you to fill in the result sets

Comment (sergeyklay) [reply phalcon/Mvc/Collection.zep]: Not all users have mongodb installed. Declaring interface here will cause fatal error for those users who don not have mongodb installed even they not use Collection class at all. Just at PHP startup time

Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: If I delete this interface, MongoDB raises an Exception

`[MongoDB\Driver\Exception\InvalidArgumentException] Class Phalcon\Test\Fixtures\Mvc\Collections\Robots does not implement MongoDB\BSON\Unserializable`

https://www.php.net/manual/en/mongodb-driver-cursor.toarray.php

File: phalcon/Mvc/Collection.zep (hunk 5 above)
Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: I made some modifications, is that good for you?

File: phalcon/Mvc/Collection.zep (hunk 7 above)
Comment (sergeyklay) [reply phalcon/Mvc/Collection.zep]: Can you declare Robots implemets Unserializable and leave Collection as is?

Comment (niden) [reply phalcon/Mvc/Collection.zep]: @

> If I delete this interface, MongoDB raises an Exception
> 
> `[MongoDB\Driver\Exception\InvalidArgumentException] Class Phalcon\Test\Fixtures\Mvc\Collections\Robots does not implement MongoDB\BSON\Unserializable`
> 
> https://www.php.net/manual/en/mongodb-driver-cursor.toarray.php

We cannot use that interface there because that introduces a limitation for the framework. Here is why:

We have introduced PDO (a few versions back). That came with some interfaces that we wanted to use to wrap Phalcon classes. As a result, because these are interfaces, we added the requirement to load PDO first before Phalcon

With v4 we introduced PSR-7 and its interfaces. New requirement: PSR has to be loaded first.

If we add this interface there, it will require all users, irrespective of whether they want to use Mongo or not, to load the mongo extension for Phalcon to work.

You might ask why does `apcu` work then if I do not have it installed? It works because we use functions for it and not interfaces. Another example is one of the `Redis` adapters. If you check the code, we create the adapter inside Phalcon and use it from there, but never expose any interfaces that Redis or any other extension uses.

I think the best way forward is to remove the interface from there and require models to implement the `Unserializable` interface for the mongo collection to work.

In short, if we introduce one interface from a different extension, Zend Engine requires it in every request, even though you might not want to use that interface ever.

Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: The solution is what then, to create an interface if it does not exist, or not to use the "typeMap" functionality ?

Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: https://www.php.net/manual/fr/mongodb.persistence.deserialization.php
This method still offers many possibilities. It allows you to transform a field into a class, for example

**Comment by CameronHall** [comment]
What is this waiting on?

File: phalcon/Mvc/Collection.zep (hunk 7 above)
Comment (sergeyklay) [reply phalcon/Mvc/Collection.zep]: Require models to implement the Unserializable interface for the MongoDB collection will do all we need here. Just  implement all interface requirements in this PR but do not declare "implement ..." interface. Indicate in the comment that the user must declare such a inheritance for each of its model of his own.

Comment (tacxticx88) [reply phalcon/Mvc/Collection.zep]: Okay, I'll fix it and take care of it.

**Comment by ruudboon** [comment]
@tacxticx88 Can you fix the 2 line enters to pass the last test?
https://travis-ci.org/phalcon/cphalcon/jobs/567264410

File: phalcon/Mvc/Collection.zep [hunk 8]
```diff
@@ -631,15 +632,15 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
...
+     *             "_id" => new \MongoDB\BSON\ObjectId("45cbc4a0e4123f6920000002"),
      *         ]
      *     ]
      * );
      *
      * echo "The robot id is ", $robot->_id, "\n";
      * ```
      */
-    public static function findFirst(array parameters = null) -> array
+    public static function findFirst(array parameters = null) -> <CollectionInterface> | null
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: This change should be reflected in the change log.

File: phalcon/Mvc/Collection.zep [hunk 9]
```diff
@@ -752,7 +753,7 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
     /**
      * Returns the value of the _id property
      *
-     * @return \MongoId
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: This change should be reflected in the change log.

File: phalcon/Mvc/Collection.zep [hunk 10]
```diff
@@ -1013,55 +1021,6 @@ abstract class Collection implements EntityInterface, CollectionInterface, Injec
         let this->skipped = skip;
     }
 
-    /**
-     * Allows to perform a summatory group for a column in the collection
-     */
-    public static function summatory(string! field, conditions = null, finalize = null) -> array
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection.zep]: Why you decided to remove this method? Anyway, should we reflect this removal in the change log?

File: phalcon/Mvc/Collection/Document.zep [hunk 11]
```diff
@@ -20,8 +20,30 @@ use Phalcon\Mvc\Collection\Exception;
  * This component allows Phalcon\Mvc\Collection to return rows without an associated entity.
  * This objects implements the ArrayAccess interface to allow access the object as object->x or array[x].
  */
-class Document implements EntityInterface, ArrayAccess
+class Document implements EntityInterface, \ArrayAccess, \Countable
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection/Document.zep]: This change should be reflected in the change log.

File: phalcon/Mvc/Collection/Manager.zep [hunk 12]
```diff
@@ -104,21 +105,21 @@ class Manager implements InjectionAwareInterface, EventsAwareInterface
     }
 
     /**
-     * Returns the connection related to a model
+     * Returns the connection related to a collection
      *
-     * @return \Mongo
+     * @return \MongoDB\Database
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection/Manager.zep]:
```suggestion
     * @return \MongoDB\Database|AdapterInterface
```

And this change should be reflected in the change log.

File: phalcon/Mvc/Collection/Manager.zep [hunk 13]
```diff
@@ -278,30 +279,30 @@ class Manager implements InjectionAwareInterface, EventsAwareInterface
          */
         let eventsManager = this->eventsManager;
         if typeof eventsManager == "object" {
-            return eventsManager->fire("model:" . eventName, model, data);
+            return eventsManager->fire("collection:" . eventName, collection, data);
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection/Manager.zep]: This change should be reflected in the change log.

File: phalcon/Mvc/Collection/ManagerInterface.zep [hunk 14]
```diff
@@ -41,56 +41,56 @@ interface ManagerInterface
...
      */
-    public function getConnection(<CollectionInterface> model) -> <AdapterInterface>;
+    public function getConnection(<CollectionInterface> collection) -> <AdapterInterface>;
 
     /**
-     * Returns a custom events manager related to a model
+     * Returns a custom events manager related to a collection
      */
-    public function getCustomEventsManager(<CollectionInterface> model) -> <EventsManagerInterface>;
+    public function getCustomEventsManager(<CollectionInterface> collection) -> var | null;
```
Comment (sergeyklay) [inline phalcon/Mvc/Collection/ManagerInterface.zep]: I propose change this to 
```suggestion
    public function getCustomEventsManager(<CollectionInterface> collection) -> <EventsManagerInterface> | null;
```

as well as https://github.com/phalcon/cphalcon/blob/3addad41e14ee83d6bb76138fb4706cadff50182/phalcon/Mvc/Collection/Manager.zep#L92

File: phalcon/Mvc/CollectionInterface.zep [hunk 15]
```diff
@@ -52,7 +52,7 @@ interface CollectionInterface
     /**
      * Allows to query the first record that match the specified conditions
      */
-    public static function findFirst(array parameters = null) -> array;
+    public static function findFirst(array parameters = null) -> <CollectionInterface> | null;
```
Comment (sergeyklay) [inline phalcon/Mvc/CollectionInterface.zep]: This change should be reflected in the change log.

File: phalcon/Mvc/CollectionInterface.zep [hunk 16]
```diff
@@ -124,9 +124,4 @@ interface CollectionInterface
      * @param mixed id
      */
     public function setId(id);
-
-    /**
-     * Check whether validation process has generated any messages
-     */
-    public function validationHasFailed() -> bool;
```
Comment (sergeyklay) [inline phalcon/Mvc/CollectionInterface.zep]: Why you decided to remove this method? I would like to see a descriptive reason in the change log

**Review by sergeyklay — COMMENTED** [review]
There are some minor questions I would like understand to. Also we'll need update change log (see my comments).  But at the moment it seems good to me.

File: phalcon/Mvc/CollectionInterface.zep (hunk 16 above)
Comment (tacxticx88) [reply phalcon/Mvc/CollectionInterface.zep]: The method did not exist in the collection, I just synchronized

**Comment by ruudboon** [comment]
@tacxticx88 If you have time can you rebase your branch to see where we stand?

**Comment by tacxticx88** [comment]
@ruudboon I'll do that. Yes, my work was moved to a private storage facility before the switch to Phalcon v4.

**Comment by ruudboon** [comment]
@tacxticx88 We're currently upgrading our testsuite as well. If you need some help rewriting the test let us know.

**Comment by ruudboon** [comment]
@tacxticx88 Any news on this? If you have time to work on this can you switch to the 4.1.x branch? thnx

**Comment by niden** [comment]
Mentioning this one more time just in case. 

For this PR to work and to be part of core, we need to ensure that it does not reference any classes or interfaces that a different extension has (in this case the php mongo extension). Doing so will require that extension to be loaded first before Phalcon even starts. Example is the PSR extension which is now a requirement for Phalcon to work.

**Comment by tacxticx88** [comment]
Hello, I'm currently looking for a solution to this problem because the Collection class should implement the interface of MongoDb to be able to automatically caste the model and its child classes..

**Comment by tacxticx88** [comment]
I suggest to finally write the ODM in PHP in order to take advantage of MongoDb's Composer repository.  In my case the work is already well advanced. What do you think of that?  MongoDb having its driver written mostly in PHP.

**Comment by niden** [comment]
> I suggest to finally write the ODM in PHP in order to take advantage of MongoDb's Composer repository. In my case the work is already well advanced. What do you think of that? MongoDb having its driver written mostly in PHP.

@tacxticx88 I think that might be the best solution at least for the time being until we regroup and figure out a way to solve the technical issues that will appear by including the mongo interfaces in the core.

We can set up an incubator repository and all the work and support can go there. People will then be able to add the component using composer.

Maybe later on we can include this functionality with FFI in the core or something similar - not sure now just thinking out loud.

I agree, we need to get this done in the incubator.

**Comment by ruudboon** [comment]
@tacxticx88 Do you agree on closing this one?

**Comment by tacxticx88** [comment]
> @tacxticx88 Do you agree on closing this one?

Yes. See the new phalcon/incobator-mongodb repository

**Comment by ruudboon** [comment]
Closing this in favour of the new https://github.com/phalcon/incubator-mongodb repo
